### PR TITLE
🛰️ Sync Push

### LINKED ISSUE #732: 🆙 Add Payload(Body) request object for POST /push/{id} method
## Description
Prepare the set of classes we'll need as **body** in  **POST /push/{id}**  method

## Dependencies
None

## Definition of done
- [ ] Prepare classes as defined in Docs and use existing API* classes as payloads if possible
- [ ] Add necessary Moshi annotations and custom adapters if needed
- [ ] Add serialization tests

### LINKED ISSUE #733: 🆙 Add response object for POST /push/{id} method
## Description
Add a set of classes for parsing response from **POST /push/{id}** method

## Dependencies
None

## Definition of done
- [ ] Prepare classes as defined in Docs for **POST /push/{id}** response
- [ ] Add necessary Moshi annotations and custom adapters if needed
- [ ] Add deserialization tests

### LINKED ISSUE #734: 🆙  Define sync POST /push/{id} method in our sync interface
## Description
Add **push** method to our sync retrofit interface and client

## Dependencies
#717 #732 #733 

## Definition of done
- [ ] Define push method in sync Retrofit interface
- [ ] Use request and response objects created in #732 and #733 
- [ ] Define push method in sync client
- [ ] Add tests if possible

### PR DESCRIPTION
**Pull request by heytherewill**
### Summary
This adds the sync push method and the things needed to serialize it

### Technical considerations
We still need to figure out a database view to fetch all data that needs syncing
For all other questions, ask @simonrozsival 😂 

### Why is it done like this and not some other way?
This is different from the Xamarin implementation, but the idea is to make things simpler

### Relationships
Closes #732
Closes #733
Closes #734

### Pull Request checklist
- [X] The code does not hinder the app's performance
- [X] The code does not introduce regressions
- [X] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests

### Pointers
Review commit by commit

### Review template
- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones

## Merge permissions
Anyone. anytime

### CODE REVIEW

File: api/src/main/java/com/toggl/api/network/models/sync/PushAction.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    val type: ActionType,
+    val meta: MetaInfo?,
+    val payload: ActionResult<T>?
+)
+
+@JsonClass(generateAdapter = true)
+data class MetaInfo(
+    val id: Long?,
+    @Json(name = "client_assigned_id")
+    val clientAssignedId: Long?,
```
Comment (simonrozsival) [inline api/src/main/java/com/toggl/api/network/models/sync/PushAction.kt:17]: This has to be a `String?` - the tests even contain the value as a string.

File: api/src/main/java/com/toggl/api/network/models/sync/PushResponse.kt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import com.toggl.api.models.SyncApiTag
+import com.toggl.api.models.SyncApiTask
+import com.toggl.api.models.SyncApiTimeEntry
+import com.toggl.api.models.SyncApiUser
+import com.toggl.api.models.SyncApiWorkspace
+
+@JsonClass(generateAdapter = true)
+data class PushResponse(
+    @Json(name = "time_entries")
+    val timeEntries: List<PushAction<SyncApiTimeEntry>>,
```
Comment (simonrozsival) [inline api/src/main/java/com/toggl/api/network/models/sync/PushResponse.kt:17]: This should be lists of `ActionResult<T>`, not `PushAction<T>`.

File: api/src/main/java/com/toggl/api/network/models/sync/PushAction.kt (hunk 1 above)
Comment (heytherewill) [reply api/src/main/java/com/toggl/api/network/models/sync/PushAction.kt:17]: Why though?Our client will only ever use longs; could it be that we get client ids from other clients? 🤔

Comment (simonrozsival) [reply api/src/main/java/com/toggl/api/network/models/sync/PushAction.kt]: How does it work then? It has to be serialized into a string and then when it's deserialized, it has to be parsed and converted from string into a long somewhere. Where is (or will) that implemented?

image

You can never get a `client_assigned_id` from any other client.

Comment (simonrozsival) [reply api/src/main/java/com/toggl/api/network/models/sync/PushAction.kt]: I just noticed commit 37853df in which you fixed it.

**Comment by heytherewill** [comment]
@daividssilverio It was addressed internally; the structure indeed is different

**Review by semanticer — APPROVED** [review]
It's looking good. And again, great job with that crazy adapter 👍

File: api/src/main/java/com/toggl/api/network/models/sync/PushResponse.kt (hunk 2 above)
Comment (simonrozsival) [reply api/src/main/java/com/toggl/api/network/models/sync/PushResponse.kt:17]: As @heytherewill pointed out, we discussed this internally, and it should actually be `PushAction<T>` (the `ActionResult<T>` is in fact the `payload` of the action).

**Review by simonrozsival — APPROVED** [review]
The `ActionResultJsonAdapter` is scary, but it seems to work just fine. I think that there are test cases for all important scenarios.

**Pull request merged**